dbus/safelauncher.go

```diff
@@ -0,0 +1,97 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package dbus
+
+import (
+	"net/url"
+	"os/exec"
+
+	"fmt"
+
+	"github.com/godbus/dbus"
+)
+
+const safeLauncherIntrospectionXML = `
+<interface name='io.snapcraft.SafeLauncher'>
+	<method name='OpenURL'>
+		<arg type='s' name='url' direction='in'/>
+	</method>
+</interface>`
+
+var (
+	allowedURLSchemes = []string{"http", "https", "mailto"}
+)
+
+// SafeLauncher implements the 'com.canonical.SafeLauncher' DBus interface
+type SafeLauncher struct{}
+
+// Name returns the name of the interface this object implements
+func (s *SafeLauncher) Name() string {
+	return "io.snapcraft.SafeLauncher"
+}
+
+// IntrospectionData gives the
+func (s *SafeLauncher) IntrospectionData() string {
+	return safeLauncherIntrospectionXML
+}
+
+func runXdgOpen(args ...string) error {
+	err := exec.Command("xdg-open", args...).Run()
```

Comment:
NOTE: On more recent versions `xdg-open` is gone (and in fact in Artful it is a broken shell script that doesn't work) and we should use `gio-open` or something like that. I think we need to detect and use the right tool for the job.